ls1intum/Artemis
# [Feature] Exercise Calculation Modes

### Checklist

### Motivation and Context



### Description

This PR is thoroughly described in the issue #2650, please have a look at this. 

TL;TR:
**We have introduced a new enum to all exercise types which allows to distinguish how the exercise is to be rated towards the overall score of a course or an exam.** 
The three enums are 
- "INCLUDED_COMPLETELY" 
  - Exercises have maximal points > 0 and bonus Points >= 0
  - The max points of the exercise counts towards the overall score, the optional bonus points can lead to a score of over 100% for a student.
- "INCLUDED_AS_BONUS"
  - Exercises have maximal points > 0 and no bonus Points
  - The complete exercise counts as bonus and does not increase the overall score. So a student can only receive positive points.
- "NOT_INCLUDED"
  - Exercises have maximal points > 0 and no bonus Points
  - The complete exercise does not count towards the overall score at all. So a student does not get negative points by not participating.

Hint: The number of points is effectively only a weighting of the task, i.e. a task with 4 points is twice as much effort/difficulty/value as a task with 2 points.
Closes #2650

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. **VERY IMPORTANT: CREATE A NEW COURSE / EXAM. THIS WILL CURRENTLY NOT WORK WITH EXERCISES CREATED BEFORE THIS PR**
3. Create an exercise of all 4 possible types (Prog., Modeling, Text, and Quiz) and set the points according to this table:

|                      | TextExercise   | ProgrammingExercise | FileUploadExercise | ModellingExercise | Points (Bonus)
|----------------------|----------------|---------------------|--------------------|-------------------|-------------------|
| INCLUDED_COMPLETELY    | COURSE \| EXAM | COURSE \| EXAM      | COURSE \| EXAM     | COURSE \| EXAM    | 1-10  (0-10)
| INCLUDED_AS_BONUS    | COURSE \| EXAM | COURSE \| EXAM      | COURSE \| EXAM     | COURSE \| EXAM    | 1-10
| NOT_INCLUDED | COURSE \| EXAM | COURSE \| EXAM      | COURSE \| EXAM     | COURSE \| EXAM    | 1-10
|                      |                |                     |                    |                   |                   |

4. Set the `IncludedInOverallScore` badge to `INCLUDED_COMPLETELY`
4.1 Verify that you can enter points for max and bonus. Max needs to be > 0 and bonus can be >= 0.
**4.2 After submitting the exercise verify that your score is displayed correctly in all views (course-exercise-row, course-scores, exam-scores)**

_**Example**:
Imagine you have a student and the following three exercises in a course
**Included 10p 10 bonus points | student achieves all 20 points, so score here 200%**
**Bonus 10P | student achieves all 10 points, so score here 100%**
**Not Included 10P | student achieves all 10p, so score here 100%** --> this will be ignored completely in the course score calculation

The "normal" points a student can achieve is 10 (10 from the included exercise), but a student can reach up to 30 points (10 from included, 10 bonus from included, 10 from pure bonus exercise). Our example student would than have 30 out of 10 points and therefore a course score of 300%_






Note for Quiz exercises:  
The complete quiz is either one of the three enums. So questions within a quiz are always of the same type. Furthermore it is allowed to give 0 points to individual questions but in total the points must be > 0. Also bonus points are not yet enabled. For this you need to create a INCLUDED_AS_BONUS quiz.

### Test Coverage

Server: 
- No relevant changes in server coverage, as the most changes were done in the client. 

Client:
- course-scores.component.ts: 95%
- exam-score-dtos.model.ts. 100%
- exam-scores.component.ts: 90%
- students-exam-import-button.component.ts: 100%
- file-upload-assessment.component.ts: 72%
- modeling-submission.component.ts: 72%
- course-statistics.component.ts: 85%
- code-editor-tutor-assessment-container.component.ts: 80%
- text-assessment-base.component.ts: 71%
- course-score-calculation.service.ts: 81%

### Screenshots


INCLUDED_COMPLETELY (with bonus):
![image](https://user-images.githubusercontent.com/29718932/104827285-50c11a00-585c-11eb-9d98-3b19365bf92f.png)



INCLUDED_AS_BONUS or NOT_INCLUDED:
![image](https://user-images.githubusercontent.com/29718932/104827256-0d66ab80-585c-11eb-9fe3-89704febf650.png)
![image](https://user-images.githubusercontent.com/29718932/104827258-122b5f80-585c-11eb-8e09-c3f6c46f83b5.png)

Course-Exercise-Row (tags):
![image](https://user-images.githubusercontent.com/29718932/104827265-1d7e8b00-585c-11eb-8f34-eef558687cce.png)


Course Statistics:
Modeling as pure bonus points (INCLUDED_AS_BONUS) exercise and programming as normal (INCLUDED_COMPLETELY):
![image](https://user-images.githubusercontent.com/29718932/104827272-3c7d1d00-585c-11eb-8a12-3a186b4f6d11.png)

Exam Mode
![Greenshot 2021-01-23 13 27 21](https://user-images.githubusercontent.com/29718932/105578820-2d79fb80-5d83-11eb-9881-5e09c1e5b2a6.png)
![Greenshot 2021-01-23 13 37 41](https://user-images.githubusercontent.com/29718932/105578811-205d0c80-5d83-11eb-874d-fc59aa3c0f17.png)
![image](https://user-images.githubusercontent.com/29718932/105607307-4d6ae880-5d9e-11eb-8fe0-4de138dcfa8f.png)
